File: week6/pset6/credit/credit.py
from cs50 import get_string

checksum = 0


def main():

    # Promt the user for an input
    number = get_string("Number: ")

    # Get first 2 digits for card type identification
    header_digits = int(number[:2])

    # Get number length
    length = len(number)

    # Apply Luhn's algorithm
    algorithm(number, length)

    # Identify card type
    identification(number, length, header_digits)


def algorithm(number, length):

    global checksum

    # Luhn's algorithm
    for i in range(1, length + 1):
        # Take every second digit starting from the end
        digit = int(number[-i])
        """ Odd index is the first digit in the credit card number and even is the second"""
        # If index is even
        if i % 2 == 0:
            # If the number has 2 digits
            if digit * 2 > 9:
                # Substracting 9 is the same as adding two digits together
                checksum += digit * 2 - 9
            else:
                checksum += digit * 2
        # If index is odd
        else:
            checksum += digit


def identification(number, length, header_digits):

    global checksum

    # Check if product after Luhn's algorithm ends with 0, if yes = True, no = False
    if checksum % 10 == 0:
        flag = True
    else:
        flag = False

    # Card identification
    if length == 15 and (header_digits == 34 or header_digits == 37) and flag:
        print("AMEX")
    elif length == 16 and 50 < header_digits < 56 and flag:
        print("MASTERCARD")
    elif int(number[:1]) == 4 and (length == 13 or length == 16) and flag:
        print("VISA")
    else:
        print("INVALID")


# Script guard
if __name__ == "__main__":
    main()

File: week1/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int count, digit, sum1, sum2;
long n, number, temp;

int main(void)
{
    // Prompt for input
    do
    {
        number = get_long("Type in credit card number\n");
    }
    while (number < 0);

    // Calculate the checksum
    // Checksum is the card authentication algorithm
    temp = number;
    while (temp > 0)
    {
        sum2 += temp % 10;
        temp /= 10;
        digit = temp % 10;
        if (digit >= 5)
        {
            sum1 += (digit * 2) % 10;
            sum1 += (digit * 2) / 10;
        }
        else
        {
            sum1 += digit * 2;
        }
        temp /= 10;
    }

    // Check for card lenght and starting digits
    // If above checks out print AMEX, VISA, MASTERCARD or INVALID
    if ((sum1 + sum2) % 10 == 0)
    {

        // Lenght check
        n = number;
        while (n > 0)
        {
            n = n / 10;
            count++;
        }

        // Sorting if AMEX, VISA, MASTERCARD or INVALID
        if (count == 15 && (number / 10000000000000 == 34 || number / 10000000000000 == 37))
        {
            printf("AMEX\n");
        }
        else if (count == 16 && (number / 100000000000000 == 51 || number / 100000000000000 == 52 || number / 100000000000000 == 53
                                 || number / 100000000000000 == 54 || number / 100000000000000 == 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((count == 13 || count == 16) && (number / 1000000000000 == 4 || number / 1000000000000000 == 4))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

}